crates/chain-state/src/chain_info.rs

```diff
@@ -112,15 +112,25 @@ impl ChainInfoTracker {
 
     /// Sets the safe header of the chain.
     pub fn set_safe(&self, header: SealedHeader) {
-        self.inner.safe_block.send_modify(|h| {
-            let _ = h.replace(header);
+        self.inner.safe_block.send_if_modified(|current_header| {
+            if current_header.as_ref() != Some(&header) {
+                let _ = current_header.replace(header);
+                return true
+            }
```

Comment:
we should just check the hash here because we expect that this is properly sealed hence checking the hash is sufficient